src/librustc_typeck/check/mod.rs

```diff
@@ -3234,7 +3236,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             // an "opportunistic" vtable resolution of any trait bounds on
             // the call. This helps coercions.
             if check_closures {
-                self.select_obligations_where_possible(false);
+                // We don't use `select_obligations_where_possible` to try to figure out if the
+                // obligation is comming from a single fn call argument, and if it is, we point
+                // at the expression corresponding to that argument, instead of the call.
+                if let Err(
```

Comment:
Would be good to not regress readability of this already-long method further by making this bit a method.